### PR TITLE
Tests: testSchema. MySQL over 8.1 (add versioned data set)

### PR DESCRIPTION
**Pull request by bscheshirwork**
I run test with `MySQL=^8`

I can see lower case in foreign key names. 

```sh
There were 5 failures:

1) yiiunit\framework\db\mysql\SchemaTest::testTableSchemaConstraints with data set "3: foreign key" ('T_constraints_3', 'foreignKeys', array(yii\db\ForeignKeyConstraint Object (...)))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    '{"columnNames":["C_fk_id_1","C_fk_id_2"],"foreignColumnNames":["C_id_1","C_id_2"],"foreignTableName":"t_constraints_2","onDelete":"CASCADE","onUpdate":"CASCADE"}' => yii\db\ForeignKeyConstraint Object (...)
+    '{"columnNames":["C_fk_id_1","C_fk_id_2"],"foreignColumnNames":["c_id_1","c_id_2"],"foreignTableName":"t_constraints_2","onDelete":"CASCADE","onUpdate":"CASCADE"}' => yii\db\ForeignKeyConstraint Object (...)
 )

/var/www/html/tests/framework/db/SchemaTest.php:732
/var/www/html/tests/framework/db/SchemaTest.php:681
/var/www/html/vendor/phpunit/phpunit/phpunit:52

2) yiiunit\framework\db\mysql\SchemaTest::testTableSchemaConstraintsWithPdoUppercase with data set "3: foreign key" ('T_constraints_3', 'foreignKeys', array(yii\db\ForeignKeyConstraint Object (...)))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    '{"columnNames":["C_fk_id_1","C_fk_id_2"],"foreignColumnNames":["C_id_1","C_id_2"],"foreignTableName":"t_constraints_2","onDelete":"CASCADE","onUpdate":"CASCADE"}' => yii\db\ForeignKeyConstraint Object (...)
+    '{"columnNames":["C_fk_id_1","C_fk_id_2"],"foreignColumnNames":["c_id_1","c_id_2"],"foreignTableName":"t_constraints_2","onDelete":"CASCADE","onUpdate":"CASCADE"}' => yii\db\ForeignKeyConstraint Object (...)
 )

/var/www/html/tests/framework/db/SchemaTest.php:732
/var/www/html/tests/framework/db/SchemaTest.php:699
/var/www/html/vendor/phpunit/phpunit/phpunit:52

3) yiiunit\framework\db\mysql\SchemaTest::testTableSchemaConstraintsWithPdoLowercase with data set "3: foreign key" ('T_constraints_3', 'foreignKeys', array(yii\db\ForeignKeyConstraint Object (...)))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    '{"columnNames":["C_fk_id_1","C_fk_id_2"],"foreignColumnNames":["C_id_1","C_id_2"],"foreignTableName":"t_constraints_2","onDelete":"CASCADE","onUpdate":"CASCADE"}' => yii\db\ForeignKeyConstraint Object (...)
+    '{"columnNames":["C_fk_id_1","C_fk_id_2"],"foreignColumnNames":["c_id_1","c_id_2"],"foreignTableName":"t_constraints_2","onDelete":"CASCADE","onUpdate":"CASCADE"}' => yii\db\ForeignKeyConstraint Object (...)
 )

/var/www/html/tests/framework/db/SchemaTest.php:732
/var/www/html/tests/framework/db/SchemaTest.php:717
/var/www/html/vendor/phpunit/phpunit/phpunit:52

```

I can see 
```php
        $connection->getSlavePdo()->setAttribute(PDO::ATTR_CASE, PDO::CASE_UPPER);
```
and 
```php
        $connection->getSlavePdo()->setAttribute(PDO::ATTR_CASE, PDO::CASE_LOWER);
```
provide lower case in constrain too.

**this attribute is not attribute about case of foreign key name? Only for column name?**

**May be this is a Error in PDO driver? Or MySQL error?** 

I can't find any info about set this behavior throw settings.

Also I can't find... So... I can find someone like. May be this is source of problem.

https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-1.html
> referenced_table_schema and referenced_table_name field values in the mysql.foreign_keys data dictionary table were not stored in lowercase when lower_case_table_names was enabled. (Bug #25495714)

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  |no
| Breaks BC?    | no
| Tests pass?   | yes
```sh
 docker-compose -f ../yii2-phpunit-docker-composition.yml run --rm --entrypoint bash php
root@2d371b5fa9dd:/var/www/html# vendor/bin/phpunit --filter testTableSchemaConstraints tests/framework/db/mysql/SchemaTest.php
PHPUnit 4.8.34 by Sebastian Bergmann and contributors.

............................................................

Time: 3.63 minutes, Memory: 6.00MB

OK (60 tests, 96 assertions)
root@2d371b5fa9dd:/var/www/html# 
```

| Fixed issues  | #15248

### CODE REVIEW

**Comment by bscheshirwork** [comment]
I try with 8.0 without path.
```sh
root@28852ab0ce9a:/var/www/html# vendor/bin/phpunit --filter testTableSchemaConstraints tests/framework/db/mysql/SchemaTest.php
PHPUnit 4.8.34 by Sebastian Bergmann and contributors.

............................................................

Time: 2.08 minutes, Memory: 6.00MB

OK (60 tests, 96 assertions)
root@28852ab0ce9a:/var/www/html# 

```
All is ok.

So. I think this is MySQL errors with 
https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-1.html
> referenced_table_schema and referenced_table_name field values in the mysql.foreign_keys data dictionary table were not stored in lowercase when lower_case_table_names was enabled. (Bug #25495714)

File: tests/framework/db/mysql/SchemaTest.php [hunk 1]
```diff
@@ -34,7 +49,13 @@ public function constraintsProvider()
         $result['3: foreign key'][2][0]->foreignTableName = new AnyCaseValue('T_constraints_2');
         $result['3: check'][2] = false;
 
+        // https://github.com/yiisoft/yii2/issues/15248
+        if (version_compare($this->getServerVersion(), '8.0.0', '>=') === true) {
```
Comment (schmunk42) [inline tests/framework/db/mysql/SchemaTest.php]: I do not really understand the underlying issue, but it looks wrong to me to add version check logic to tests, just to make them pass. Although I can not provide another solution yet.

Comment (bscheshirwork) [reply tests/framework/db/mysql/SchemaTest.php]: I think we have error in MySQL. We can't change this(

**Comment by bscheshirwork** [comment]
http://bugs.mysql.com/88718

**Comment by schmunk42** [comment]
> referenced_table_schema and referenced_table_name field values in the mysql.foreign_keys data dictionary table were not stored in lowercase when lower_case_table_names was enabled. (Bug #25495714)

Isn't it like this: MySQL fixed a bug in 8.0.1. Before that `lower_case_table_names` had no effect on foreign keys. We could check for that *setting* on the server and test accordingly. So there could be tests for lower-case only and mixed case, which is converted to lower-case, if the setting is enabled and if the server is 8.0.1 or above.

Not sure if this is easier to understand, but it would be also interesting how the other MySQL implementations handle this.

**Comment by bscheshirwork** [comment]
@schmunk42 I think this fix tell about part of foreign key constrain, who depends on table name?

**Comment by bscheshirwork** [comment]
i.e. something like this
```sql
CREATE TABLE `T_constraints_3`
(
    `C_id` INT NOT NULL,
    `C_fk_id_1` INT NOT NULL,
    `C_fk_id_2` INT NOT NULL,
    CONSTRAINT `CN_constraints_3` FOREIGN KEY (`C_fk_id_1`, `C_fk_id_2`) REFERENCES `T_constraints_2` (`C_id_1`, `C_id_2`) ON DELETE CASCADE ON UPDATE CASCADE
)
ENGINE = 'InnoDB' DEFAULT CHARSET = 'utf8';
```
=>
```sql
CREATE TABLE `T_constraints_3`
(
    `C_id` INT NOT NULL,
    `C_fk_id_1` INT NOT NULL,
    `C_fk_id_2` INT NOT NULL,
    CONSTRAINT `CN_constraints_3` FOREIGN KEY (`C_fk_id_1`, `C_fk_id_2`) REFERENCES `t_constraints_2` (`C_id_1`, `C_id_2`) ON DELETE CASCADE ON UPDATE CASCADE
)
ENGINE = 'InnoDB' DEFAULT CHARSET = 'utf8';
```
`T_constraints_2` => `t_constraints_2`

**Comment by bscheshirwork** [comment]
@schmunk42 and again:
> 
I can see

        $connection->getSlavePdo()->setAttribute(PDO::ATTR_CASE, PDO::CASE_UPPER);

and

        $connection->getSlavePdo()->setAttribute(PDO::ATTR_CASE, PDO::CASE_LOWER);

provide lower case in constrain too.

**Comment by samdark** [comment]
@bscheshirwork do you want to re-do it as @schmunk42 suggested?

**Comment by terabytesoftw** [comment]
The tests in MySQL 8.1 work perfectly, so I think this issue can be closed.

**Comment by bizley** [comment]
Agree, closing.

**Pull request closed without merging**